Implement dynamic prefix for subapps

achimnol opened the pull request:
## What do these changes do?

This allows use of variables (regular expressions) in the subapp prefixes.

## Are there changes in behavior for the user?

Existing codes are not affected.

## Related issue number

I am making this PR after discussion with @asvetlov in gitter.

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [x] Documentation reflects the changes
- [x] I'm already in `CONTRIBUTORS.txt`
- [x] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` for example (588.bugfix)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

achimnol committed:
Implement dynamic prefix for subapps

achimnol commented:
My implementation approach:
 * Instead of calling `add_prefix()` to all resources in the subapp, I strip the matched prefix from the request to resolve the subapp's routes, cloning the request with a new URL.
    - This would incur some performance overheads, but I tried to implement it "correctly" first.
    - Existing codes that use plain subapp prefixes won't be affected.
 * User-defined web handlers still receive the original request object, so I belive that user codes should not be affected in any way.

achimnol committed:
Fix import sorting errors

achimnol committed:
Add changes.

achimnol committed:
Update docs.

achimnol committed:
Fix spelling error.

achimnol committed:
Fix a logical error to mis-add prefixes to subapp's resources.

* This behavior should be consistent with the constructor.

achimnol committed:
Fortify tests to cover nested subapps with mixed plain/dynamic prefixes.

achimnol commented:
To allow access to all overriden variables, we could use `ChainMap` / `MultiDict` instead of `dict` for building `UrlMappingMatchInfo` class, but I'm not sure if just enumerating multi-values in the nesting order would be sufficient or not.
@asvetlov How about your ideas?

achimnol committed:
Allow retrieval of overriden variables using ChainMap.

achimnol commented:
Okay, I implemented using ChainMap. :wink:

achimnol committed:
Fix spelling error.

achimnol committed:
Make the index of match_info.maps consistent with match_info._apps

* So let it add an empty dict when nesting.

asvetlov commented on tests/test_urldispatch.py:
```diff
@@ -1003,7 +1005,7 @@ def test_url_for_in_resource_route(router):
 
 def test_subapp_get_info(app, loop):
     subapp = web.Application()
-    resource = subapp.add_subapp('/pre', subapp)
```
Good catch!

asvetlov commented:
@achimnol thanks for PR.
It covers 99% but I see the only one problem: what if `{name}` is present in both dynamic rules: once in subapp prefix itself and other one in routed sub application?

I see the solution:
1. Add `.parent` property to `UrlMappingMatchInfo`. For nested apps it will point on subapp prefix route.
2. Don't derive `UrlMappingMatchInfo` from dict but implement `collections.abc.Mapping` interface and `__getitem__` method. The method can look into `.parent` if the key is absent. 

Right now I don't care about performance, clean API is the first target. Later we can add Cythonized version of `UrlMappingMatchInfo` if needed.

What do you think about?

achimnol commented:
1. My PR handles such cases using `ChainMap` (see the test case `test_nested_dynamic_subapp_resolution`) -- `.maps` can serve the purpose of `.parent` you suggest, like the `_apps` internal property. I think `.parent` may not be very useful because from the subapp's perspective it is anyway "unknown" how many it should traverse recursively to get the right "parent". However, if you prefer this way, I can rewrite it along with 2.

2. The original code was deriving `UrlMappingMatchInfo` from `dict` so I followed that style. If you'd like to change this, I can rewrite it.

achimnol commented:
To change `UIrlMappingMatchInfo` to be a self-nested data structure, I think we need to rewrite its `apps` property handling as well because it also treats the list of nested apps as its own property and this clashes with self-nesting `match_info` objects. Also I still don't get the benefit of having per-app-level `match_info` objects with duplicate routes since sub-app resources are not treated as final destinations. 😕

achimnol committed:
Apply review comments partially

* Let UrlMappingMatchInfo inherit collections.abc.Mapping instead of
  ChainMap directly.

  - Make the order of inheritance consistent with UrlDispatcher.

* Make the order of the nested dictionaries returned by "variable_maps"
  consistent with the order of the nested apps returned by "apps".

asvetlov commented:
I've convinced. Let's use chained map approach.
But please wait for my careful review.

asvetlov commented on aiohttp/web_urldispatcher.py:
```diff
@@ -172,12 +172,13 @@ def url_for(self, *args, **kwargs):
 
 
-class UrlMappingMatchInfo(dict, AbstractMatchInfo):
+class UrlMappingMatchInfo(AbstractMatchInfo, Mapping):
```
I really like the fact that `UrlMappingMatchInfo` is immutable now.

asvetlov commented on aiohttp/web_urldispatcher.py:
```diff
@@ -172,12 +172,13 @@ def url_for(self, *args, **kwargs):
 
     def __init__(self, match_dict, route):
-        super().__init__(match_dict)
+        super().__init__()
```
`super()` call is not needed, Python machinery will work fine without it.

asvetlov commented on aiohttp/web_urldispatcher.py:
```diff
@@ -230,8 +231,24 @@ def set_current_app(self, app):
+        return tuple(reversed(self._variables.maps))
+
+    def add_variables(self, match_dict):
+        self._variables.maps.append(match_dict)
```
I suspect you should insert a new dict *in front* of maps -- like we do it for `add_app()`.
Moreover maybe merging `add_app()` and `add_variables()` functionality makes sense.
What do you think about `_add_app(app, match_dict)` method as replacement for `add_app`.
Adding underscore emphasizes that the method is private. `add_app()` was never a part of documented public API, I think the change is safe in backward compatibility perspective.

achimnol replied:
`ChainMap` resolves from the beginning of its `maps`. The point here is that `add_variable()` is called *after* resolving the subapp's routes, which means that subapp's match variables are added to the chain map first. Since we want to give priority to the innermost subapp's variables in the merged view of this chain map, this `append` is natural.

I think merging this method into `add_app(app, match_dict)` is a good idea. :+1:

achimnol replied:
Hm.... merging into a single method causes a lot of problem:
* When instantiating `UrlMappingMatchInfo` we already have `match_dict` (so the chain map begins with a single layer) but `_app` is None and `_apps` is an empty tuple. This makes the number of nested maps and apps inconsistent.
* Changing `UrlMappingMatchInfo`'s constructor to accept an explicit `app` argument requires `Route` and `Resource` classes to keep track of their owner `app` because `resolve()` method should be able to indicate the exact `app` instance to create the match info object. `request.app` is a self null reference there.

The root cause of this mess is that `app` and `match_dict` are added to `UrlMappingMatchInfo` at different places and timings. :confused:

asvetlov commented on aiohttp/web_urldispatcher.py:
```diff
@@ -230,8 +231,24 @@ def set_current_app(self, app):
+
     def __repr__(self):
-        return "<MatchInfo {}: {}>".format(super().__repr__(), self._route)
+        return "<MatchInfo {}: {}>".format(dict(self).__repr__(), self._route)
```
`repr(self._variables)` instead of `repr(dict(self))` is better I believe. There is no reason to hide chained variables.

achimnol replied:
This was to make existing test cases that compares the `repr` result to pass. Then I will change those test cases to match with `ChainMap`'s `repr` result.

asvetlov commented on aiohttp/web_urldispatcher.py:
```diff
@@ -230,8 +231,24 @@ def set_current_app(self, app):
         self._frozen = True
 
+    @property
+    def variable_maps(self):
```
Maybe let's use just `maps` name?
`maps` is shorted, in documentation we could refer to `ChainMap.maps`. `maps` is mentally more familiar to people if they know about `ChainMap` already.

achimnol replied:
This is also related to the resolution order of the chain map and the build order of it is reversed.
Also I wanted to make it read-only since `ChainMap`'s `maps` property is mutable.

asvetlov commented on docs/web_reference.rst:
```diff
@@ -2181,8 +2189,8 @@ In general the result may be any object derived from
 
-   Inherited from :class:`dict` and :class:`AbstractMatchInfo`. Dict
-   items are filled by matching info and is :term:`resource`\-specific.
+   Inherited from :class:`collections.ChainMap` and :class:`AbstractMatchInfo`.
```
The doc is not relevant anymore, `Mapping` should replace `ChainMap`.
Also `.variable_maps` or `.maps` attribute should be documented.

asvetlov reviewed (commented):
Sorry for delay and thank you for patience.

I have some comments.
Let's discuss my proposals.

achimnol committed:
Apply review comments.

* Kept add_map() (formerly add_variables) and add_app() methods
  sepearte since they are often called in disjoint contexts where the
  other is out-of-scope.

* Renamed variable_maps to maps.

* Updated the docs related to add_subapp(), MatchInfo, and
  DynamicSubAppResource.

asvetlov commented on aiohttp/web_urldispatcher.py:
```diff
@@ -227,11 +227,29 @@ def set_current_app(self, app):
+    def add_map(self, match_dict):
+        if self._frozen:
+            raise RuntimeError("Cannot change maps stack after .freeze() call")
+        self._variables.maps.append(match_dict)
```
Let's consider the following:
```
async def handler(request):
    return web.Response(text=request.math_info['name'])

subapp = web.Application()
subapp.add_routes([web.get('/{name}', handler)])

app = web.Application()
app.add_subapp('/{name}', subapp)
```
What `name` should be returned by the handler?

achimnol replied:
In this case, `request.match_info['name']` in the handler will return the matched value of the innermost `{name}` definition: the route one.

asvetlov replied:
Perfect!

asvetlov commented on aiohttp/web_urldispatcher.py:
```diff
@@ -674,6 +693,57 @@ def __repr__(self):
+        self._app = app
+        self._prefix = prefix
+
+    def url_for(self, *args, **kwargs):
```
I see more complex problem.
`.url_for()` for all resources from sub-application are broken for dynamic prefixed resource.

Sorry, but the problem is blocker for merging.

achimnol replied:
Yes, that's what I mentioned in the gitter channel.
I think rewriting `url_for()` to have a recursive architecture so that different types of resources could be chained require a lot of work, so I'd like to suggest it as a separate PR.
Maybe we could mark this PR as "provisional" if merged before it.

achimnol replied:
I'm currently very busy due to my business work -- once I get some time after a few weeks, I'll give a look on the extension plan for `url_for()`.

asvetlov replied:
The problem is that without `url_for` fix not only `url_for()` for subapp itself doesn't work but all `url_for()` calls for subapp resources are broken.
That's why I think everything should be done in single PR.

achimnol committed:
Implement url_for() with dynamic subapp prefix.

* Changed add_prefix() to accept resource instances instead of prefix
  strings.

* url_for() now recursively call the parent's url_for() and prefix
  the current url with it.
  For this, url_for() with PrefixedSubAppResoucre/DynamicSubAppResource
  are no longer errors.

* Changed PrefixedSubAppResource.resolve() to also clone the request
  with the stripped URL like DymaicSubAppResource.resolve().

* This implementation does not break any existing code. :)

* TODO: I will add some test cases and extra checks to the code.

achimnol commented:
Implemented support for `url_for()`.
There is an issue to decide on: unlike `match_info`, the `**kwargs` passed to `url_for()` does not distinguish which key is for which subapp prefix. @asvetlov Do you have an idea for this?

achimnol commented:
Ah, since dynamic prefix is not a released API yet, I think it is safe to make the proposed name-and-predixed-keys scheme *optional*.
If the user wants to avoid key collision then they can use this optional name prefixes, but when not, they can just use as-is. No breaking for existing codes. Yay.

asvetlov commented on aiohttp/web_urldispatcher.py:
```diff
@@ -637,27 +669,26 @@ class PrefixedSubAppResource(PrefixResource):
 
     async def resolve(self, request):
-        if not request.url.raw_path.startswith(self._prefix):
+        if not request.url.path.startswith(self._prefix):
```
Why not `raw_path`?

achimnol replied:
What if the prefix includes arbitrary unicode characters?

asvetlov replied:
Mixing PCT encoded and decoded data can make a mess easy.
I believe we should follow the strategy:
1. Convert all user provided *path*, *prefix* etc into *normalized* (PCT-encoded actually) form by using `yarl`. Already normalized paths passed *as-is* by yarl.
2. Normalize given HTTP path (we do it).
3. Match normalized path to already pre-normalized patterns.
4. Return `request.match_info['name']` in **PCT-decoded** form because users most likely want getting `이름` instead of `'%EC%9D%B4%EB%A6%84`

If aiohttp code violates the rule somewhere -- it should be fixed.
Do you agree?

achimnol replied:
Okay, I agree to keep the rule consistent. Let me inspect and fix up this PR.

achimnol replied:
I think then `PlainResource.__init__()` also need this, like `PrefixResource`:
```python
self._path = URL.build(path=path).raw_path
```

achimnol committed:
Use prefixed variable keys in match_dict and url_for for dynamic subapps

* Fixed some unimplemented/mistaken parts of the previous commit
* Reverted back to use 'raw_path' where it had been used before
  as it still works as desired.

achimnol commented:
Now if we have a nested dynamic subapp chain:
```python
root_app = web.Application()
outer_app = web.Application()
inner_app = web.Application()
resource = inner_app.router.add_resource('/{var}')
route = resource.add_route('GET', ...)
outer_app.add_subapp('/{var}', inner_app, name='b')
root_app.add_subapp('/{var}', outer_app, name='a')
```

Then `match_dict` after resolving `"/abc/def/ghi"` has the content:
```python
{
  'a.var': 'abc',
  'b.var': 'def',
  'var': 'ghi',
}
```

`url_for(**{'a.var': '123', 'b.var': '456', 'var': '789'})` gives:
```python
'/123/456/789'
```

No more ChainMap!

achimnol committed:
Add more tests and fix variable scoping when collided

achimnol committed:
More test

achimnol commented:
Rebased against latest master.

achimnol committed:
Update docs.

achimnol committed:
Remove docs for deprecated reverse-ChainMap attribute

achimnol commented:
I think it's ready to get reviewed. :)

achimnol committed:
Revert my changes related to PCT encoded paths and fix PlainResource

* All paths (even regexs against them) stored in resources are already
  encoded.  They are only decoded when passed back to the user.

* PlainResource was missing this explicit encoding when initialized.

achimnol committed:
Fix flake8 error

achimnol committed:
Add/modify tests to use Unicode and fix duplicate add_prefix() calls

asvetlov commented:
Let me meditate for a while. 
``"name."`` looks... strange. How to manage a situation if `name` argument for `app.add_subapp` was not passed? 
Another thing worries me is: suppose I have a common-usable application like admin or debug toolbar.
I'm adding it with variable prefix and the subapp becomes broken: it has a knowledge about its required parents for generating urls. 
The worse thing is that breakage appears on subapp usage (url generations for filing HTTP templates), not on subapp adding.

If you have ideas how to make it steady -- please share your vision. 
We a free to invite a new API or even restrict subapp type to `VariablePrefixAwareSubapp` inherited from `web.Application` but the game rules should be discussed and fixed in stone before the PR merging.

I understand your striving to make subapp prefix flexible but my guts feel that we should be very careful with chosen design. Otherwise fitting a opened can of worms will be painful.

I suggest excluding the PR from aiohttp 3.1 but working hard to publish it as part of aiohttp 3.2
Your idea for chaining `match_info` is brilliant.
There is a very visible request for providing chained info for nested applications: 
- #2689 for accessing parent configuration like `request.app['db']`
- #2314 for building URLs from sibling sub-apps
We should make a uniform approach for all of these.
 
P.S.
I'll go to Ho Chi Minh tomorrow evening for a couple weeks. Not sure if I will have enough time to work on aiohttp these dates.
The city is much closer to Seoul than Kiev but unfortunately still so far for meeting with you offline :(

achimnol commented:
I see. You want more isolation between apps -- and I agree with that.
I thought having explicit names was better than nested scopes because it is clear to read.

I'm fine to go back to the first ChainMap design for `match_info`.
The problem is `url_for()`.

To allow use `url_for()` in the innermost ("leaf") app *without* knowing whatever variables are included in its outer subapp prefixes, then I think we need a strong assumption that `url_for()` is only used inside request handlers where `match_info` is available in the current task-local context.
Once `url_for()` has access to the nested `match_info`'s ChainMap then it can fill the prefix variables using the matched values of the current request's path.

Still we have two issues:

1. Python asyncio currently lacks task-local context support which automatically propagates to subtasks. To workaround this, it may be necessary to change the `url_for()` API to accept the `request` object (since it plays the role of handler-local context now) or its `match_info` attribute. But this would break use of `url_for()` in template rendering.
2. `url_for()` may be used to refer URLs with prefixes other than the current request's path has. In this case users need to provide the prefix variables anyway, like my "name." design (of course the concrete method may be different).

@asvetlov What do you think?

achimnol commented:
More thoughts:
Maybe could we ingore the 2nd case if we stick with "isolated" apps?
If each app never calls `url_for()` against routes in other apps, then filling the prefix variables with the current request's path should be ok.

achimnol commented:
More more thoughts:
When composing multiple apps, cross-reference to URLs in other apps may happen in a specific direction only: from outer apps to inner apps, not to parent/sibling apps.

From the perspective of inner apps, it's safe to assume that the aiohttp's web framework will take care of filling outer prefix variables with the handler-local match_info context.

From the perspective of outer apps, they need to explicitly give the value of perfix variables when calling `url_for()` of routes of inner apps. In this specific case, we need a way to distinguish the duplicate variable names in the prefix names.

achimnol committed:
Add extra check to prevent having multiple prefixes

asvetlov commented:
Maybe name prefixes are not bad but I need a time to experiment with.
Regarding to mandatory params from outer apps etc:
1. We can invite a new API for it, something like `request.build_url()` (the name is a subject for changing)
2. Context local variables (implicit context) are not required as we always have *explicit context* as `request`.
3. Inner/outer applications can be useful terminology.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/aio-libs/aiohttp?pullRequest=2756) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/aio-libs/aiohttp?pullRequest=2756) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/aio-libs/aiohttp?pullRequest=2756) it.</sub>

Dreamsorcerer commented:
@achimnol Do we want to revive this PR? I suspect it'll need some adjustments since we overhauled the routing algorithm sometime ago.
